### PR TITLE
Enable GitHub code scanning with CodeQL (to replace the soon-to-be-deprecated LGTM.com)

### PR DESCRIPTION
**Pull request by sj**
Hi team!

You might have seen that LGTM.com will go off-line [later this month](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/). The team that built LGTM (including me!) [joined GitHub](https://github.blog/2019-09-18-github-welcomes-semmle/) a couple of years ago, and we've natively integrated the CodeQL engine that powers LGTM into GitHub.

This PR enables the CodeQL analysis for three.js. Just like with LGTM, alerts will show up in pull requests (but better). In addition, all alerts can be browsed on the "Security" tab of the repository.

I tried to configure GitHub code scanning as similarly/closely as possible to the original LGTM configuration — do feel free to tweak and adjust as you see fit. If you have any questions, check out the FAQ below, and I'm of course happy to help as well. 

### FAQ
<details>
<summary>Click here to expand the FAQ section</summary>

#### How often will the code scanning analysis run?
By default, code scanning will trigger a scan with the CodeQL engine on the following events:
On every pull request — to flag up potential security problems for you to investigate before merging a PR.
On every push to your default branch and other protected branches — this keeps the analysis results on your repository’s *Security* tab up to date.
Once a week at a fixed time — to make sure you benefit from the latest updated security analysis even when no code was committed or PRs were opened.

#### What will this cost?
Nothing! The CodeQL engine will run inside GitHub Actions, making use of your [unlimited free compute minutes for public repositories](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#about-billing-for-github-actions).

#### What types of problems does CodeQL find?
The CodeQL engine that powers GitHub code scanning is the exact same engine that powers LGTM.com. The exact set of rules has been tweaked slightly, but you should see almost exactly the same types of alerts as you were used to on LGTM.com: we’ve enabled the [`security-and-quality` query suite](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#using-queries-in-ql-packs) for you.

#### How do I upgrade my CodeQL engine?
No need! New versions of the CodeQL analysis are constantly deployed on GitHub.com; your repository will automatically benefit from the most recently released version.

#### The analysis doesn’t seem to be working
If you get an error in GitHub Actions that indicates that CodeQL wasn’t able to analyze your code, please [follow the instructions here](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/troubleshooting-the-codeql-workflow) to debug the analysis.

#### How do I disable LGTM.com?
If you have LGTM’s automatic pull request analysis enabled, then you can [follow these steps to disable the LGTM pull request analysis](https://lgtm.com/help/lgtm/managing-automated-code-review#disabling-pr-integration). You don’t actually need to remove your repository from LGTM.com; it will automatically be removed in the next few months as part of the deprecation of LGTM.com ([more info here](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/)).

#### Which source code hosting platforms does code scanning support?
GitHub code scanning is deeply integrated within GitHub itself. If you’d like to scan source code that is hosted elsewhere, we suggest that you create a mirror of that code on GitHub.

#### How do I know this PR is legitimate?
This PR is filed by the official LGTM.com GitHub App, in line with the [deprecation timeline that was announced on the official GitHub Blog](https://github.blog/2022-08-15-the-next-step-for-lgtm-com-github-code-scanning/). The proposed GitHub Action workflow uses the [official open source GitHub CodeQL Action](https://github.com/github/codeql-action). If you have any other questions or concerns, please join the discussion [here](https://github.com/orgs/community/discussions/29534) in the official GitHub community!

#### I have another question / how do I get in touch?
Please join the discussion [here](https://github.com/orgs/community/discussions/29534) to ask further questions and send us suggestions!

</details>

### CODE REVIEW

**Comment by sj** [comment]
As you can see, the analysis is working: it's identified some alert and flagged them up as *new* (because they didn't exist before this PR, and it therefore blames this PR for 'introducing' them). These are all similar to the alerts that LGTM identified before.

The CodeQL analysis at PR time will of course only flag up new alerts — just like LGTM did.

Hope you find this PR helpful!

File: .github/workflows/codeql-code-scanning.yml
Comment (LeviPesin) [inline .github/workflows/codeql-code-scanning.yml:33]: Maybe we can use pinning for the actions (i.e. using the commit hash instead of the version)? Otherwise this can potentially leave the repository open to a supply chain attack in actions, with which the repository's secrets could be stealed, (not sure if there are any enabled in three.js).

Comment (sj) [reply .github/workflows/codeql-code-scanning.yml:33]: If you trust the provider of the Action (in this case: GitHub) and its ability to put measures in place to prevent malicious take-overs, then it's fine to pin to a `@v2` tag. But if you want to be super secure, then you are indeed very welcome to [pin to the Action's commit SHA](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions). You could then even [use Dependabot to help keep your Action SHAs up to date!](https://docs.github.com/en/code-security/dependabot/working-with-dependabot/keeping-your-actions-up-to-date-with-dependabot).

I'll leave it up to the three.js team to decide what to do here. Maintainers should feel free to edit the contents of the PR :slightly_smiling_face:

File: .github/codeql-config.yml
Comment (mrdoob) [inline .github/codeql-config.yml:8]: I don't think we need this line? 🤔

File: .github/codeql-config.yml
Comment (Mugen87) [inline .github/codeql-config.yml:4]: This line can also be removed now.

File: .github/codeql-config.yml
Comment (Mugen87) [inline .github/codeql-config.yml:6]: And this one, too.

File: .github/codeql-config.yml
Comment (marcofugaro) [inline .github/codeql-config.yml:1]: shouldn't this go under [`on.push.paths-ignore`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore) of the `codeql-code-scanning.yml` action?

Comment (sj) [reply .github/codeql-config.yml:1]: This particular setting tells the CodeQL analysis knows which files not to analyse. This means that regardless of when the analysis runs (e.g. on a PR, or on a push, or on schedule), CodeQL won't bother analysing these files or flagging up any alerts for these paths. I've based that on [the existing `.lgtm.yml` configuration file in the repository](https://github.com/mrdoob/three.js/blob/5368e6a206eba8afd93760a4b8c815e13a43c257/.lgtm.yml).

The [on.push.paths-ignore](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore) is a setting that stops an Action (like the CodeQL analysis) from being triggered if changes to a _particular_ path are _pushed_. It can be useful to apply that in a CodeQL context, but it would mean that the CodeQL analysis would still flag up alerts when it is triggered by, for example, a PR being opened.

File: .github/codeql-config.yml
Comment (sj) [reply .github/codeql-config.yml]: Without this line, you'll get an [extraordinarily large number of alerts](https://github.com/mrdoob/three.js/pull/25056/checks?check_run_id=9844725516) in the `manual/3rdparty/monaco-editor` subtree. The CodeQL analysis in LGTM automatically excluded minified files ([example](https://github.com/mrdoob/three.js/blob/dev/manual/3rdparty/monaco-editor/min/vs/basic-languages/markdown/markdown.js)), but it doesn't do that by default on GitHub.

But of course it's up to you! If you're interested in seeing these alerts and fixing them, then that's great :rocket:

**Review by Mugen87 — APPROVED** [review]

**Pull request merged**
